compat.c

```diff
@@ -4,37 +4,508 @@
+					   const char *, bool);
+	int (*iio_channel_attr_write_longlong)(const struct iio_channel *,
+					       const char *, long long);
+	int (*iio_channel_attr_write_double)(const struct iio_channel *,
+					     const char *, double);
+	unsigned int (*iio_channel_get_type)(const struct iio_channel *);
+	unsigned int (*iio_channel_get_modifier)(const struct iio_channel *);
+	long (*iio_channel_get_index)(const struct iio_channel *);
+	const struct iio_data_format *
+		(*iio_channel_get_data_format)(const struct iio_channel *);
+
+	void (*iio_channel_convert)(const struct iio_channel *,
+				    void *, const void *);
+	void (*iio_channel_convert_inverse)(const struct iio_channel *,
+					    void *, const void *);
+
+	void (*iio_channel_enable)(const struct iio_channel *,
+				   struct iio_channels_mask *);
+	void (*iio_channel_disable)(const struct iio_channel *,
+				    struct iio_channels_mask *);
+	bool (*iio_channel_is_enabled)(const struct iio_channel *,
+				       const struct iio_channels_mask *);
+
+	/* Channel masks */
+	struct iio_channels_mask * (*iio_create_channels_mask)(unsigned int);
+	void (*iio_channels_mask_destroy)(struct iio_channels_mask *);
+
+	/* Buffers */
+	struct iio_buffer * (*iio_device_create_buffer)(const struct iio_device *,
+							unsigned int,
+							const struct iio_channels_mask *);
+	void (*iio_buffer_destroy)(struct iio_buffer *);
+	void (*iio_buffer_cancel)(struct iio_buffer *);
+	int (*iio_buffer_enable)(struct iio_buffer *);
+	int (*iio_buffer_disable)(struct iio_buffer *);
+
+	const struct iio_device * (*iio_buffer_get_device)(const struct iio_buffer *);
+	const struct iio_channels_mask *
+		(*iio_buffer_get_channels_mask)(const struct iio_buffer *);
+	void (*iio_buffer_set_data)(struct iio_buffer *, void *);
+	void * (*iio_buffer_get_data)(const struct iio_buffer *);
+
+	/* Blocks */
+	struct iio_block * (*iio_buffer_create_block)(struct iio_buffer *, size_t);
+	void (*iio_block_destroy)(struct iio_block *);
+	struct iio_buffer * (*iio_block_get_buffer)(const struct iio_block *);
+	int (*iio_block_enqueue)(struct iio_block *, size_t, bool);
+	int (*iio_block_dequeue)(struct iio_block *, bool);
+	void * (*iio_block_start)(const struct iio_block *);
+	void * (*iio_block_first)(const struct iio_block *,
+				  const struct iio_channel *);
+	void * (*iio_block_end)(const struct iio_block *);
+	ssize_t (*iio_block_foreach_sample)(const struct iio_block *,
+					    const struct iio_channels_mask *,
+					    ssize_t (*callback)(const struct iio_channel *,
+								void *, size_t, void *),
+					    void *);
+
+	/* Misc */
+	size_t (*iio_strlcpy)(char * __restrict, const char * __restrict, size_t);
+	char *(*iio_strdup)(const char *);
+	void (*iio_strerror)(int, char *, size_t);
+	void (*iio_prm_printf)(const struct iio_context_params *params,
+			       unsigned int msg_level,
+			       const char *fmt, ...);
+};
+
+static struct compat *compat_lib_ptr;
+
+#define IIO_CALL(fn) (*compat_lib_ptr->fn)
```

Comment:
There is a reason why it's done this way: `__VA_ARGS__` must have at least one argument.

So this:
`#define IIO_CALL(fn, ...) (*compat_lib_ptr->fn)(__VA_ARGS__)`

would make it impossible to call like this:
`IIO_CALL(iio_get_backends_count);`

iio-debug.h has the same problem and gets around it with some black magic. But in this case it seemed smarter to define `IIO_CALL` like that.

Reply:
Ahh I remember I also had to do some magic in the adrv9002 stream example... I guess this is because we cannot tak gnu extensions as granted? IIRC, with the extension we don't have this constrain.

Reply:
Yes, for instance Visual Studio does not have the GNU extensions ;)